src/reconcile/rec_child.c

```diff
@@ -15,90 +15,80 @@
+        return (0);
     }
 
     /*
-     * Deal with any underlying disk blocks.
-     *
-     * First, check to see if there is an address associated with this leaf: if there isn't, we're
-     * done, the underlying page is already gone. If the page still exists, check for any
-     * transactions in the system that might want to see the page's state before it's deleted.
-     *
-     * If any such transactions exist, we cannot discard the underlying leaf page to the block
-     * manager because the transaction may eventually read it. However, this write might be part of
-     * a checkpoint, and should we recover to that checkpoint, we'll need to delete the leaf page,
-     * else we'd leak it. The solution is to write a proxy cell on the internal page ensuring the
-     * leaf page is eventually discarded.
+     * A visible entry can be in a prepared state and checkpoints skip in-progress prepared changes.
+     * We can't race here, the entry won't be visible to the checkpoint, or will be in a prepared
+     * state, one or the other.
      *
-     * If no such transactions exist, we can discard the leaf page to the block manager and no cell
-     * needs to be written at all. We do this outside of the underlying tracking routines because
-     * this action is permanent and irrevocable. (Clearing the address means we've lost track of the
-     * disk address in a permanent way. This is safe because there's no path to reading the leaf
-     * page again: if there's ever a read into this part of the name space again, the cache read
-     * function instantiates an entirely new page.)
+     * We should never see an in-progress prepare in eviction: when we check to see if an internal
+     * page can be evicted, we check for an unresolved fast-truncate, which includes a fast-truncate
+     * in a prepared state, so it's an error to see that during eviction.
      */
-    if (ref->addr != NULL && !__wt_page_del_active(session, ref, true)) {
-        WT_RET(__wt_ref_block_free(session, ref));
+    WT_ORDERED_READ(prepare_state, page_del->prepare_state);
+    if (prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED) {
+        WT_ASSERT(session, !F_ISSET(r, WT_REC_EVICT));
 
-        /* Any fast-truncate information can be freed as soon as the delete is stable. */
-        __wt_overwrite_and_free(session, ref->ft_info.del);
+        *statep = WT_CHILD_ORIGINAL;
+        return (0);
     }
 
     /*
-     * If the original page is gone, we can skip the slot on the internal page.
+     * Deal with underlying disk blocks. If there are readers that might want to see the page's
+     * state before it's deleted, or the fast-delete can be undone by RTS, we can't discard the
+     * pages. Write a cell to the internal page with information describing the fast-delete.
      */
-    if (ref->addr == NULL) {
-        *statep = WT_CHILD_IGNORE;
+    if (__wt_page_del_active(session, ref, true) ||
```

Comment:
Not related to this code, but the `__wt_page_del_active` internally uses the `page_del->timestamp` for the visible_all scenario also, it must be `page_del->durable_timestamp`

Reply:
Yes, it is a data inconsistency error. I will create another ticket for it.